### DISCUSSION TITLE
From Static Site with RSS feed (astro) to fedi?

### DISCUSSION THREAD

**Discussion by reggi** [fedify-dev/fedify#37] [Ideas]
I have a project with [Github](https://github.com/astoria-tech/meetup), [Blog](https://astoria.app/), [RSS](https://astoria.app/rss.xml), released as a static site. What I'm interested in is creating a simple single-page fedi app like in the deno deploy demo, but that fetches the rss once a day using deno cron (or has a authed `/fetch-update` endpoint my SSG github action can trigger)  and provides an activity pub feed from a static blog. 

I know that Astro (the framework i'm using for SSG) has SSG and a new database implementation, I'm not as interested in using SSR for my blog and I'm ok with splitting the sites into two apps astoria.app (on gh pages) and fedi.astoria.app (deno deploy). However, I can imagine an implementation using Astro's new DB (similar to the deno one0 would provide a lot of value for the astro community as well. 

I'll dig through the source to see if this is possible, but just curious if anyone else has had this idea, if it exists already, or if there's something preventing me from doing this.

**Comment by dahlia** [comment]
The idea you're describing sounds like it could be easily implemented with Fedify. You could [integrate Fedify with a micro-framework like Hono][1] to implement a */fetch-update* endpoint, or as you said, you could use Deno Cron to implement it. When the event is triggered, you could convert the update to a [`Note`] or [`Article`] object, wrap it in a [`Create`] activity, and [send it to the followers using `Context.sendActivity()` method][2].

However, there are of course easier ways to do this using tools that already exist, such as [Hatsu].

Additionally, you might find the following series of articles by Maho Pacheco helpful: [A guide to implement ActivityPub in a static site (or any website)](https://maho.dev/2024/02/a-guide-to-implement-activitypub-in-a-static-site-or-any-website/).

[1]: https://fedify.dev/manual/integration#hono
[`Note`]: https://jsr.io/@fedify/fedify/doc/~/Note
[`Article`]: https://jsr.io/@fedify/fedify/doc/~/Article
[`Create`]: https://jsr.io/@fedify/fedify/doc/~/Create
[2]: https://fedify.dev/manual/send
[Hatsu]: https://hatsu.cli.rs/

**Reply by reggi** [reply]
I made a first pass here https://github.com/reggi/rss-fedi but couldn't figure out why I couldn't see posts I think I was missing the `sendActivity` method. 👍  Wow thanks, I didn't know about Hatsu, it seems nice, and I like rust too but ideally:

```ts
import { rssFedi } from 'jsr:xxx'

await Deno.serve(rssFedi({
  handle: "meetup",
  title: "Meetup Events",
  desc: "Upcoming events from the Astoria Tech Meetup",
  published: "2022-04-13",
  url: "https://astoria.app",
  name: "Astoria Tech Meetup",
  summary: "Upcoming events from the Astoria Tech Meetup",
})
```

**Reply by dahlia** [reply]
FYI you can [acquire a `Context` instance using `Federation.getContext()` method][1].

[1]: https://fedify.dev/manual/context#where-to-get-a-context-object

**Reply by reggi** [reply]
I implmented this here https://github.com/reggi/rss-fedi/blob/main/fedi.ts#L36-L59 but nothing showing on the account https://indieweb.social/@meetup@fedi.deno.dev, any thoughts? I also didn't see any logging within fedify, might be useful to know what's going on.

**Reply by dahlia** [reply]
Could you show me the content of the `Create` instance?

**Reply by reggi** [reply]
@dahlia 
this line calls `federation.createContext` https://github.com/reggi/rss-fedi/blob/main/fedi.ts#L44C1-L44C56
this line calls `new Federation` https://github.com/reggi/rss-fedi/blob/main/fedi.ts#L68

Is this what you're referring to?

**Reply by dahlia** [reply]
If you could log the `Create` instance [here](https://github.com/reggi/rss-fedi/blob/54417e7e6c603f6651e4a5fbfb817f50d687978a/fedi.ts#L49-L54), it would be easier for me to help you debug.

**Reply by reggi** [reply]
I just cloned the blog and got it running on its own and I can see posts 😻 🥳 

https://indieweb.social/@Example@thomasreggi-fedi-blog-87.deno.dev
https://thomasreggi-fedi-blog-87.deno.dev/posts
https://github.com/reggi/fedi-blog

I'm just going to slowly remove fresh and try and make a single jsr package that does this:

```
Deno.serve(rssFedi({
   rssFeed: "https://astoria.app"
}))
```

**Comment by reesericci** [comment]
Just thinking out loud here - would it be possible for a direct integration with Astro through the use of JS routes, like serverless functions w/the Netlify or Vercel adapters?

**Reply by dahlia** [reply]
We don't officially support Astro integration yet, but it should be possible since Astro has [middleware].  See the [integration manual] for more information.

[middleware]: https://docs.astro.build/en/guides/middleware/
[integration manual]: https://fedify.dev/manual/integration

**Reply by dahlia** [reply]
In the next release, v0.12.0, Fedify will have [Astro integration](https://unstable.fedify.dev/manual/integration#astro)!

**Comment by dahlia** [comment]
We now have [Astro integration](https://fedify.dev/manual/integration#astro). See also [*Building a federated blog* tutorial](https://fedify.dev/tutorial/astro-blog) (which utilizes Astro).
